File: backend/app/models/notification.py
from datetime import datetime
from enum import Enum
from typing import Optional

from sqlalchemy import Boolean, DateTime, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base


class NotificationType(str, Enum):
    LEAD_CREATED = "lead_created"
    LEAD_ASSIGNED = "lead_assigned"
    LEAD_STATUS_CHANGED = "lead_status_changed"
    PRODUCT_CREATED = "product_created"
    PRODUCT_UPDATED = "product_updated"
    PRODUCT_DELETED = "product_deleted"
    APPOINTMENT_BOOKED = "appointment_booked"
    CALL_REPORT_GENERATED = "call_report_generated"


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    message: Mapped[str] = mapped_column(Text, nullable=False)
    type: Mapped[str] = mapped_column(String(50), nullable=False)
    is_read: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False, index=True)
    related_lead_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True, index=True)
    related_call_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True, index=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )


class NotificationPreference(Base):
    __tablename__ = "notification_preferences"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    notification_type: Mapped[str] = mapped_column(String(50), nullable=False)
    enabled: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: frontend/src/types/notification.ts
export type NotificationType =
    | 'lead_created'
    | 'lead_assigned'
    | 'lead_status_changed'
    | 'product_created'
    | 'product_updated'
    | 'product_deleted'
    | 'appointment_booked'
    | 'call_report_generated';

export interface Notification {
    id: number;
    user_id: number;
    message: string;
    type: NotificationType;
    is_read: boolean;
    related_lead_id?: number | null;
    related_call_id?: number | null;
    created_at: string;
}

export interface NotificationPreferenceItem {
    notification_type: NotificationType;
    enabled: boolean;
}

export interface NotificationPreferencesResponse {
    items: NotificationPreferenceItem[];
}

export interface NotificationListResponse {
    notifications: Notification[];
    total: number;
    page: number;
    page_size: number;
}
